lighthouse-core/gather/fetcher.js

```diff
@@ -0,0 +1,177 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* global document */
+
+class Fetcher {
+  /**
+   * @param {import('./driver.js')} driver
+   */
+  constructor(driver) {
+    this.driver = driver;
+    /** @type {Map<string, (event: LH.Crdp.Fetch.RequestPausedEvent) => void>} */
+    this._onRequestPausedHandlers = new Map();
+    this._onRequestPaused = this._onRequestPaused.bind(this);
+  }
+
+  /**
+   * The Fetch domain accepts patterns for controlling what requests are intercepted, but we
+   * enable the domain for all patterns and filter events at a lower level to support multiple
+   * concurrent usages. Reasons for this:
+   *
+   * 1) only one set of patterns may be applied for the entire domain.
+   * 2) every request that matches the patterns are paused and only resumes when certain Fetch
+   *    commands are sent. So a listener of the `Fetch.requestPaused` event must either handle
+   *    the requests it cares about, or explicitly allow them to continue.
+   * 3) if multiple commands to continue the same request are sent, protocol errors occur.
+   *
+   * So instead we have one global `Fetch.enable` / `Fetch.requestPaused` pair, and allow specific
+   * urls to be intercepted via `driver.setOnRequestPausedHandler`.
+   */
+  async enableRequestInterception() {
+    await this.driver.sendCommand('Fetch.enable', {
+      patterns: [{requestStage: 'Request'}, {requestStage: 'Response'}],
+    });
+    await this.driver.on('Fetch.requestPaused', this._onRequestPaused);
+  }
+
+  /**
+   * @param {string} url
+   * @param {(event: LH.Crdp.Fetch.RequestPausedEvent) => void} handler
+   */
+  async setOnRequestPausedHandler(url, handler) {
+    this._onRequestPausedHandlers.set(url, handler);
+  }
+
+  /**
+   * @param {LH.Crdp.Fetch.RequestPausedEvent} event
+   */
+  async _onRequestPaused(event) {
+    const handler = this._onRequestPausedHandlers.get(event.request.url);
+    if (handler) {
+      await handler(event);
+    } else {
+      // Nothing cares about this URL, so continue.
+      await this.driver.sendCommand('Fetch.continueRequest', {requestId: event.requestId});
+    }
+  }
+
+  async disableRequestInterception() {
+    await this.driver.off('Fetch.requestPaused', this._onRequestPaused);
+    await this.driver.sendCommand('Fetch.disable');
+    this._onRequestPausedHandlers.clear();
+  }
+
+  /**
+   * Requires that `driver.enableRequestInterception` has been called.
+   *
+   * Fetches any resource in a way that circumvents CORS.
+   *
+   * @param {string} url
+   * @param {number} timeoutInMs
+   * @return {Promise<string>}
+   */
+  async fetchResource(url, timeoutInMs = 500) {
+    if (!this.driver.isDomainEnabled('Fetch')) {
```

Comment:
should we encapsulate this inside `fetchResource` and do an idempotent `enable` + check for onRequestPausedHandlers to see if we disable when we're done?

it seems mostly like an annoying impl detail to keep track of this enabling from the perspective of a gatherer that just wanted to fetch a single resource

Reply:
I'd be cool with that.

it seems the gather-runner needs to call disable at the end of gathering (which would do nothing if no gatherer called fetchResource)?

Reply:
@patrickhulce is this what you intend? having gather runner "clean up" by calling disable so that no gather must do it themselves?

Reply:
well do we even need gatherRunner to cleanup? :)

if we just call enable/disable before and after and only actually invoke when there's no more outstanding, that would work?

if it wouldn't work for some reason, having individual gatherers call these doesn't seem that different from it happening in gather runner IMO

Reply:
also if we are going to be replacing this with a simple "Page.fetchTheThing" protocol method soon, then the whole enabling/disabling domain thing from outside is going to need to go away so I definitely vote to keep that internal to this class if possible :)